tests/e2e/ppnd-e2e/ppnd_test.go

```diff
@@ -0,0 +1,367 @@
+
+		CreateFailedPipelineRollout(failedPipelineRolloutName, Namespace, failedPipelineSpec)
+		VerifyPipelineFailed(Namespace, failedPipelineRolloutName)
+
+		time.Sleep(5 * time.Second)
+
+		// update ISBService to have data loss update
+		Document("Updating ISBService to cause a PPND change")
+		updatedISBServiceSpec := isbServiceSpec
+		updatedISBServiceSpec.JetStream.Version = initialJetstreamVersion
+
+		// need to update function
+		// update would normally cause data loss
+		UpdateISBServiceRollout(isbServiceRolloutName, failedPipelineRolloutName, updatedISBServiceSpec, func(retrievedISBServiceSpec numaflowv1.InterStepBufferServiceSpec) bool {
+			return retrievedISBServiceSpec.JetStream.Version == initialJetstreamVersion
+		}, true, false, false)
+
+		DeletePipelineRollout(failedPipelineRolloutName)
+
+	})
+
+	It("Should update a NumaflowControler even if the Pipeline is failed", func() {
+
+		// add bad edge to automatically fail Pipeline
+		failedPipelineSpec := initialPipelineSpec
+		failedPipelineSpec.Edges = append(failedPipelineSpec.Edges, numaflowv1.Edge{From: "not", To: "valid"})
+
+		CreateFailedPipelineRollout(failedPipelineRolloutName, Namespace, failedPipelineSpec)
+		VerifyPipelineFailed(Namespace, failedPipelineRolloutName)
+
+		time.Sleep(5 * time.Second)
+
+		Document("Updating Numaflow controller to cause a PPND change")
+		UpdateNumaflowControllerRollout(updatedNumaflowControllerVersion, initialNumaflowControllerVersion, failedPipelineRolloutName, true)
+
+		DeletePipelineRollout(failedPipelineRolloutName)
+
+	})
+
+	It("Should delete all remaining rollout objects", func() {
+		DeleteISBServiceRollout(isbServiceRolloutName)
+		DeleteNumaflowControllerRollout()
+	})
+
+})
+
+func createSlowPipelineRollout() {
+
+	Document("Creating a slow pipeline")
+	slowPipelineSpec = updatedPipelineSpec.DeepCopy()
+	highRPU := int64(10000000)
+	readBatchSize := uint64(1)
+	pauseGracePeriodSeconds := int64(3600) // numaflow will try to pause for 1 hour if we let it
+	slowPipelineSpec.Lifecycle.PauseGracePeriodSeconds = &pauseGracePeriodSeconds
+	slowPipelineSpec.Limits = &numaflowv1.PipelineLimits{ReadBatchSize: &readBatchSize}
+	slowPipelineSpec.Vertices[0].Source.Generator.RPU = &highRPU
+	slowPipelineSpec.Vertices[1].UDF = &numaflowv1.UDF{Container: &numaflowv1.Container{
+		Image: "quay.io/numaio/numaflow-go/map-slow-cat:stable",
+	}}
+
+	CreatePipelineRollout(slowPipelineRolloutName, Namespace, *slowPipelineSpec)
+
+	Document("Verifying that the slow pipeline was created")
+	VerifyPipelineSpec(Namespace, slowPipelineRolloutName, func(retrievedPipelineSpec numaflowv1.PipelineSpec) bool {
+		return len(slowPipelineSpec.Vertices) == len(retrievedPipelineSpec.Vertices)
+	})
+
+	VerifyPipelineRunning(Namespace, slowPipelineRolloutName, false)
+	VerifyInProgressStrategy(slowPipelineRolloutName, apiv1.UpgradeStrategyNoOp)
+
+}
+func completeSlowPipelineRolloutTest() {
```

Comment:
I know you're trying to consolidate as much shared functionality as possible into a single function, but I still think functions should be named to have some kind of meaning, which represents what it does. 

I think it would be slightly less consolidated, but better to divide it into 3, e.g.:
1. verifySlowPipelinePausing()
2. allowDataLoss()
3. DeletePipelineRollout()

Reply:
Sure! Yeah that would be a better idea, I was thinking more about consolidating the logic overall.